app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+
+        $newCategory = $request->get('category');
+        $newSiteId = $post->site_id;
+
+        if($post->site_id == 0 || $oldSiteId != $request->get('site_id')) {
+            $newSiteId = $request->get('site_id');
+        }
+
+        try {
+            if($this->currentUser->is(User::CONTRIBUTOR)) {
+                throw new BreakException();
+            }
+
+            if ($submitType == 'publish' && !Post::isUniquePostByTitle($post, ['site_id' => $newSiteId])) {
+                throw new TitleDuplicatedException('このタイトルは存在しています。公開できないのでタイトルを変更してください。');
+            }
+
+            if ($oldSiteId != $newSiteId && $oldSiteId != 0 && $oldTargetPostId != 0) {
+                $apiUrlOldTarget = ($oldSiteId != 0) ? Site::find($oldSiteId)->api_url : null;
+                try {
+                    (new CdsApiDriver($apiUrlOldTarget, ['post_id' => $oldTargetPostId]))->send('delete')->getResult();
+                } catch (\Exception $e) {
+                    if ($e->getMessage() != 'DELETE_POST_NOT_FOUND') {
+                        $this->activityLog->setInfo('Delete old post on target site failed');
+                        $this->activityLog->addMeta('ChangeTargetSiteFailed', "Delete old post on target site failed");
+                        throw $e;
+                    }
+                    $this->activityLog->addMeta('Warning', "Target post id #{$oldTargetPostId} not found");
+                }
+
+                $post->target_post_id = 0;
+            }
+
+            if (($submitType == 'publish' || ($oldStatus == 1 && $submitType == 'draft'))
+                && (!$this->currentUser->is(User::CONTRIBUTOR) && $post->site_id > 0)) {
+
+                $action = ($oldSiteId != $newSiteId) || $oldSiteId == 0
+                || ($post->published_status == 0 && $post->target_post_id == 0) ? 'publish' : 'edit';
+
+                $apiUrl = Site::find($newSiteId)->api_url;
+
+                $returnData = (new CdsApiDriver($apiUrl, [
+                    'content' => $postContent,
+                    'title' => $postTitle,
+                    'categories' => [$newCategory],
+                    'post_id' => $post->target_post_id,
+                    'post_excerpt' => $postDescription,
+                    'post_status' => $submitType,
+                    'feature_img' => $post->feature_img,
+                    'author_name' => !empty($author = User::find($post->user_id)) ? $author->username : '',
+                    'cback_url' => action('Webapp\\PostController@getEdit', $post->id),
+                    'old_post_id' => $oldPostId,
+                    'priority' => $post->priority,
+                ]))->send($action)->getResult();
+                if ($returnData->data->post_id == 0) {
+                    $this->activityLog->setInfo('Post to target site failed for unknown reason');
+                    $this->activityLog->addMeta('FailedPostToTargetSite','unknown reason');
+                    throw new \Exception();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS112)